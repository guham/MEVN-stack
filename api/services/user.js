const { User } = require('../models');

exports.findOrCreate = async (userSub) => {
  let user = await User.findOne({ sub: userSub }).exec();
  if (!user) {
    user = new User({ sub: userSub });
    await user.save();
  }
  return user;
};

exports.storeUserRefreshToken = async (userSub, refreshToken) => {
  const user = await exports.findOrCreate(userSub);
  const timestamp = new Date().getTime();
  user.refreshTokens.set(timestamp.toString(), refreshToken);
  await user.save();
  return user;
};

/**
 * @param {object} user
 * @param {string} refreshToken
 * @returns {boolean} true if the refresh token existed and has been removed, otherwise false
 */
exports.removeUserRefreshToken = async (user, refreshToken) => {
  const tokens = user.refreshTokens;
  const entry = Array.from(tokens).filter(element => element[1] === refreshToken).shift();
  if (!entry) {
    return false;
  }

  tokens.delete(entry[0]);
  await user.save();
  return true;
};
