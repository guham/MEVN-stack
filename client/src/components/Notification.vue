<!-- eslint-disable max-len -->
<template>
  <div class="fixed pin-b pin-x bg-white p-4 text-grey-darkest border-t border-grey-light">
    <div class="flex w-full justify-between items-center">
      <div class="hidden lg:flex" />
      <div class="flex mr-6">
        <span
          :class="{ 'text-red': notification.type === 'error' }"
          class="mr-2"
        >
          {{ $t(notification.type) }}
        </span>{{ $t(notification.text) }}
      </div>
      <div class="flex self-start">
        <button
          class="flex items-center px-3 py-2 text-grey-dark"
          aria-label="Close notification"
          @click="removeNotification()"
        >
          <svg
            class="fill-current h-4 w-4"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 8.586L2.929 1.515 1.515 2.929 8.586 10l-7.071 7.071 1.414 1.414L10 11.414l7.071 7.071 1.414-1.414L11.414 10l7.071-7.071-1.414-1.414L10 8.586z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</template>

<i18n>
{
  "en": {
    "error": "Error",
    "%SOMETHING_WENT_WRONG%": "Something went wrong. Please try again."
  },
  "fr": {
    "error": "Erreur",
    "%SOMETHING_WENT_WRONG%": "Une erreur est survenue. Merci de réessayer."
  },
  "es": {
    "error": "Error",
    "%SOMETHING_WENT_WRONG%": "Ha ocurrido un error. Por favor, inténtalo de nuevo."
  }
}
</i18n>

<script>
export default {
  name: 'Notification',
  props: {
    notification: {
      type: Object,
      required: true,
    },
  },

  methods: {
    removeNotification() {
      this.$emit('removeNotification');
    },
  },
};
</script>
