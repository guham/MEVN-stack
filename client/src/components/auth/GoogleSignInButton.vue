<template>
  <button
    v-if="isAuthenticated"
    id="btn-signout"
    class="btn btn-grey"
    type="button"
    @click="signOutAndRedirect()">
    Sign Out
  </button>
  <button
    v-else
    id="btn-signin"
    class="btn btn-green"
    type="button"
    @click="signIn()">
    Sign In
  </button>
</template>

<script>
import { createNamespacedHelpers } from 'vuex';

const { mapGetters, mapActions } = createNamespacedHelpers('user');

export default {
  name: 'GoogleSignInButton',

  computed: {
    ...mapGetters([
      'isAuthenticated',
    ]),
  },

  methods: {
    ...mapActions([
      'signIn',
      'signOut',
    ]),
    async signOutAndRedirect() {
      await this.signOut();
      this.$router.replace({ name: 'Home' });
    },
  },
};
</script>
