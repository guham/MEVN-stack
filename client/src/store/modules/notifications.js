import * as types from '@/store/mutation-types';

const state = {
  notification: {},
};

const getters = {
  hasNotification: state => Object.keys(state.notification).length > 0,
};

const mutations = {
  [types.ADD_NOTIFICATION](state, notification) {
    state.notification = notification;
  },

  [types.REMOVE_NOTIFICATION](state) {
    state.notification = {};
  },
};

const actions = {
  addThenRemoveNotification: {
    root: true,
    handler({ dispatch, commit }, notification) {
      notification.timer = setTimeout(() => {
        dispatch('removeNotification');
      }, 8000);
      commit(types.ADD_NOTIFICATION, notification);
    },
  },
  removeNotification({ commit, state }) {
    clearTimeout(state.notification.timer);
    commit(types.REMOVE_NOTIFICATION);
  },
};

export default {
  namespaced: true,
  state,
  getters,
  mutations,
  actions,
};
