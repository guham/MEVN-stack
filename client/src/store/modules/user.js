import authClient from '@/api/auth';
import * as types from '@/store/mutation-types';
import Notification from '@/models/notification';

const state = {
  isAuthenticated: false,
  userIsSigningIn: false,
  jwt: localStorage.getItem('token'),
  jwtExpiration: parseInt(localStorage.getItem('tokenExpiration'), 10) || null,
};

const getters = {
  isAuthenticated: state => state.isAuthenticated,
  userIsSigningIn: state => state.userIsSigningIn,
  token: state => state.jwt,
};

const mutations = {
  [types.SET_USER_AUTHENTICATED](state) {
    state.isAuthenticated = true;
  },

  [types.SET_USER_IS_SIGNING_IN](state, flag = true) {
    state.userIsSigningIn = flag;
  },

  [types.SET_USER_UNAUTHENTICATED](state) {
    state.isAuthenticated = false;
    state.jwt = null;
    state.jwtExpiration = null;
    localStorage.removeItem('token');
    localStorage.removeItem('tokenExpiration');
  },

  [types.SET_JWT](state, token) {
    state.jwt = token;
    localStorage.setItem('token', token);
  },

  [types.SET_JWT_EXPIRATION](state, tokenExpiration) {
    state.jwtExpiration = tokenExpiration;
    localStorage.setItem('tokenExpiration', tokenExpiration);
  },
};

const actions = {
  async signIn({ dispatch, commit }) {
    commit(types.SET_USER_IS_SIGNING_IN, true);

    try {
      const auth2 = window.gapi.auth2.getAuthInstance();
      const googleUser = await auth2.signIn();
      const idToken = googleUser.getAuthResponse().id_token;
      const response = await authClient.sendIdToken(idToken);
      commit(types.SET_USER_AUTHENTICATED);
      commit(types.SET_JWT, response.token);
      commit(types.SET_JWT_EXPIRATION, response.tokenExpiration);
    } catch (e) {
      // display a notification only when error happened with the API
      // auth2.signIn() will returns an object containing an *error* property if an error happened
      if (!e.error) {
        dispatch('addThenRemoveNotification', new Notification('error', '%SOMETHING_WENT_WRONG%'), { root: true });
      }
      await dispatch('signOut');
    }
    commit(types.SET_USER_IS_SIGNING_IN, false);
  },
  async signOut({ dispatch, commit }) {
    const auth2 = window.gapi.auth2.getAuthInstance();
    await auth2.signOut().catch(() => {
      dispatch('addThenRemoveNotification', new Notification('error', '%SOMETHING_WENT_WRONG%'), { root: true });
    });
    commit(types.SET_USER_UNAUTHENTICATED);
  },
};

export default {
  namespaced: true,
  state,
  getters,
  mutations,
  actions,
};
