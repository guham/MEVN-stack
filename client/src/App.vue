<!-- eslint-disable max-len -->
<template>
  <div id="app">
    <nav
      v-click-outside="closeMenu"
      class="flex items-center bg-white border-b border-grey-lighter justify-between flex-wrap fixed pin-t pin-x z-10 p-3 lg:p-4">
      <div class="flex items-center flex-no-shrink text-white mr-6">
        <img
          src="./assets/logo.png"
          class="h-8 w-8 mr-2"
          alt="Client logo">
        <router-link
          to="/"
          class="font-semibold text-xl tracking-tight text-grey-darker">Client</router-link>
      </div>
      <div class="flex block lg:hidden">
        <div
          :id="[ menuIsVisible ? 'menu-close' : 'menu-open' ]">
          <button
            class="flex items-center px-3 py-2 border rounded text-green border-green hover:text-green-lighter hover:border-green-light"
            aria-label="Toggle navigation"
            @click="toogleMenu()">
            <svg
              class="fill-current h-3 w-3"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg">
              <path
                v-if="!menuIsVisible"
                d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z"/>
              <path
                v-else
                d="M10 8.586L2.929 1.515 1.515 2.929 8.586 10l-7.071 7.071 1.414 1.414L10 11.414l7.071 7.071 1.414-1.414L11.414 10l7.071-7.071-1.414-1.414L10 8.586z"/>
            </svg>
          </button>
        </div>
      </div>
      <div
        :class="{ hidden: !menuIsVisible }"
        class="w-full block flex-grow lg:flex lg:items-center lg:w-auto">
        <div class="text-sm lg:flex-grow text-left lg:text-center px-2">
          <router-link
            to="/"
            class="link">{{ $t('home') }}</router-link>
          <template v-if="isAuthenticated">
            <span class="hidden lg:inline">|</span>
            <router-link
              to="/test"
              class="link">Test</router-link>
            <span class="hidden lg:inline">|</span>
            <router-link
              to="/foo"
              class="link">Foo</router-link>
          </template>
        </div>
        <div class="text-sm text-left lg:text-center px-2 lg:px-0">
          <router-link
            to="/parameters"
            class="link">{{ $t('parameters') }}</router-link>
          <google-sign-in-button
            class="ml-2 lg:ml-0 mt-4 lg:mt-0"
            @signIn="closeMenu()"/>
        </div>
      </div>
    </nav>
    <div class="min-h-screen pt-24 px-6 pb-6">
      <router-view/>
    </div>
  </div>
</template>

<i18n>
{
  "en": {
    "home": "Home",
    "parameters": "Parameters"
  },
  "fr": {
    "home": "Accueil",
    "parameters": "Paramètres"
  },
  "es": {
    "home": "Inicio",
    "parameters": "Parámetros"
  }
}
</i18n>

<script>
import { createNamespacedHelpers } from 'vuex';
import GoogleSignInButton from '@/components/auth/GoogleSignInButton.vue';
/* eslint-disable-next-line */
import logo from '@/assets/logo.png';

const { mapGetters } = createNamespacedHelpers('user');

export default {
  metaInfo: {
    titleTemplate: '%s - MEVN-stack',
    title: 'Client',
    meta: [
      { name: 'Description', content: 'MEVN stack' },
      { property: 'og:title', content: 'MEVN stack - client' },
      { property: 'og:type', content: 'website' },
      { property: 'og:image', content: `${process.env.VUE_APP_CLIENT_URL}${logo}` },
      { property: 'og:url', content: process.env.VUE_APP_CLIENT_URL },
      { property: 'og:description', content: 'MEVN stack' },
    ],
  },
  components: {
    GoogleSignInButton,
  },
  data: () => ({
    menuIsVisible: false,
  }),

  computed: {
    ...mapGetters([
      'isAuthenticated',
    ]),
  },

  watch: {
    $route(to, from) {
      from.name && this.closeMenu();
    },
  },

  methods: {
    toogleMenu() {
      this.menuIsVisible = !this.menuIsVisible;
    },
    closeMenu() {
      this.menuIsVisible = false;
    },
  },
};
</script>

<style>
#app {
  font-family: 'Avenir', Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #2c3e50;
}
a {
  color: #1a4731;
}
a.router-link-exact-active {
  color: #1f9d55;
}
</style>
