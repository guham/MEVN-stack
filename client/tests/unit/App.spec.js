import Vuex from 'vuex';
import App from '@/App.vue';
import GoogleSignInButton from '@/components/auth/GoogleSignInButton.vue';
import NotificationComponent from '@/components/Notification.vue';
import user from '@/store/modules/user';
import notifications from '@/store/modules/notifications';
import Notification from '@/models/notification';
import factory from './factory';

const notificationsModule = {
  namespaced: notifications.namespaced,
  state: notifications.state,
  getters: notifications.getters,
  actions: notifications.actions,
};

const userUnauthenticatedStore = new Vuex.Store({
  modules: {
    user: {
      namespaced: user.namespaced,
      getters: {
        isAuthenticated: () => false,
        userIsSigningIn: () => false,
      },
    },
    notifications: notificationsModule,
  },
});

const userAuthenticatedStore = new Vuex.Store({
  modules: {
    user: {
      namespaced: user.namespaced,
      getters: {
        isAuthenticated: () => true,
        userIsSigningIn: () => false,
      },
    },
    notifications: notificationsModule,
  },
});

describe('App.vue', () => {
  test('#app is define', () => {
    const wrapper = factory(App, userUnauthenticatedStore);
    expect(wrapper.find('app')).toBeDefined();
  });

  test('should display "Home" and "Parameters" links when the user is not authenticated', () => {
    const wrapper = factory(App, userUnauthenticatedStore);
    const template = wrapper.html();
    const links = wrapper.findAll('routerlinkwrapper-stub');
    expect(links).toHaveLength(3);
    expect(links.at(0).props().to).toBe('/');
    expect(links.at(1).props().to).toBe('/');
    expect(links.at(2).props().to).toBe('/parameters');
    expect(template).toMatchSnapshot();
  });

  test('should display restricted links when the user is authenticated', () => {
    const wrapper = factory(App, userAuthenticatedStore);
    const template = wrapper.html();
    const links = wrapper.findAll('routerlinkwrapper-stub');
    expect(links.filter(link => link.props().to === '/test').isVisible()).toBeTruthy();
    expect(links.filter(link => link.props().to === '/foos').isVisible()).toBeTruthy();
    expect(template).toMatchSnapshot();
  });

  test('toggle menu when open/close button is clicked', (done) => {
    const wrapper = factory(App, userUnauthenticatedStore);
    wrapper.vm.$router.push('/');
    expect(wrapper.vm.menuIsVisible).toBeFalsy();
    const button = wrapper.find('#menu-open > button');
    expect(button).toBeDefined();
    button.trigger('click');
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.menuIsVisible).toBeTruthy();
      done();
    });
  });

  test('should collapse the menu on route change', (done) => {
    const wrapper = factory(App, userUnauthenticatedStore);
    wrapper.vm.$router.push('/');
    // user opened the menu
    wrapper.vm.menuIsVisible = true;
    expect(wrapper.find('#menu-close').isVisible()).toBeTruthy();
    // user clicked on parameters link
    wrapper.vm.$router.push('/parameters');
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.menuIsVisible).toBeFalsy();
      done();
    });
  });

  test('calls `closeMenu()` when @signIn happens', (done) => {
    const wrapper = factory(App, userUnauthenticatedStore);
    const spyCloseMenu = jest.spyOn(wrapper.vm, 'closeMenu');
    wrapper.find(GoogleSignInButton).vm.$emit('signIn');
    wrapper.vm.$nextTick(() => {
      expect(spyCloseMenu).toHaveBeenCalled();
      done();
    });
  });

  test('should display a notification when one is added to the Notifications store', (done) => {
    const notificationStore = new Vuex.Store({
      modules: {
        user,
        notifications: {
          ...notificationsModule,
          ...{ state: { notification: new Notification() } },
        },
      },
    });

    const wrapper = factory(App, notificationStore);
    wrapper.vm.$nextTick(() => {
      expect(wrapper.find(NotificationComponent).isVisible()).toBeTruthy();
      done();
    });
  });

  test('calls store action `removeNotification` when @removeNotification happens', (done) => {
    const actions = {
      removeNotification: jest.fn(),
    };
    const notificationStore = new Vuex.Store({
      modules: {
        user,
        notifications: {
          ...notificationsModule,
          ...{ state: { notification: new Notification() } },
          ...{ actions },
        },
      },
    });

    const wrapper = factory(App, notificationStore);
    wrapper.find(NotificationComponent).vm.$emit('removeNotification');
    wrapper.vm.$nextTick(() => {
      expect(actions.removeNotification).toHaveBeenCalled();
      done();
    });
  });
});
