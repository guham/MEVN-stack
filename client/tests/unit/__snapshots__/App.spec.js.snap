// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue should display only "Home" link when the user is not authenticated 1`] = `
<div id="app">
  <nav class="flex items-center bg-white border-b border-grey-lighter justify-between flex-wrap fixed pin-t pin-x z-10 p-3 lg:p-4">
    <div class="flex items-center flex-no-shrink text-white mr-6"><img src="./assets/logo.png" alt="Client logo" class="h-8 w-8 mr-2">
      <router-link-stub to="/" tag="a" event="click" class="font-semibold text-xl tracking-tight text-grey-darker">Client</router-link-stub>
    </div>
    <div class="flex block lg:hidden">
      <div id="menu-open"><button class="flex items-center px-3 py-2 border rounded text-green border-green hover:text-green-lighter hover:border-green-light"><svg viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg" class="fill-current h-3 w-3">
            <path d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z"></path>
          </svg></button></div>
    </div>
    <div class="w-full block flex-grow lg:flex lg:items-center lg:w-auto hidden">
      <div class="text-sm lg:flex-grow text-left lg:text-center px-2">
        <router-link-stub to="/" tag="a" event="click" class="link">home</router-link-stub>
        <!---->
      </div>
      <div class="text-sm text-left lg:text-center px-2 lg:px-0">
        <router-link-stub to="/parameters" tag="a" event="click" class="link">parameters</router-link-stub>
        <googlesigninbutton-stub class="ml-2 lg:ml-0 mt-4 lg:mt-0"></googlesigninbutton-stub>
      </div>
    </div>
  </nav>
  <div class="pt-24 p-6">
    <router-view-stub name="default"></router-view-stub>
  </div>
</div>
`;

exports[`App.vue should display restricted links when the user is authenticated 1`] = `
<div id="app">
  <nav class="flex items-center bg-white border-b border-grey-lighter justify-between flex-wrap fixed pin-t pin-x z-10 p-3 lg:p-4">
    <div class="flex items-center flex-no-shrink text-white mr-6"><img src="./assets/logo.png" alt="Client logo" class="h-8 w-8 mr-2">
      <router-link-stub to="/" tag="a" event="click" class="font-semibold text-xl tracking-tight text-grey-darker">Client</router-link-stub>
    </div>
    <div class="flex block lg:hidden">
      <div id="menu-open"><button class="flex items-center px-3 py-2 border rounded text-green border-green hover:text-green-lighter hover:border-green-light"><svg viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg" class="fill-current h-3 w-3">
            <path d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z"></path>
          </svg></button></div>
    </div>
    <div class="w-full block flex-grow lg:flex lg:items-center lg:w-auto hidden">
      <div class="text-sm lg:flex-grow text-left lg:text-center px-2">
        <router-link-stub to="/" tag="a" event="click" class="link">home</router-link-stub> <span class="hidden lg:inline">|</span>
        <router-link-stub to="/test" tag="a" event="click" class="link">Test</router-link-stub> <span class="hidden lg:inline">|</span>
        <router-link-stub to="/foo" tag="a" event="click" class="link">Foo</router-link-stub>
      </div>
      <div class="text-sm text-left lg:text-center px-2 lg:px-0">
        <router-link-stub to="/parameters" tag="a" event="click" class="link">parameters</router-link-stub>
        <googlesigninbutton-stub class="ml-2 lg:ml-0 mt-4 lg:mt-0"></googlesigninbutton-stub>
      </div>
    </div>
  </nav>
  <div class="pt-24 p-6">
    <router-view-stub name="default"></router-view-stub>
  </div>
</div>
`;
